File: web/src/utils/apiError.ts
// Copyright (c) 2026 ZyvorAI Labs Private Limited. All rights reserved.
// Proprietary software — see LICENSE in the repository root.
// https://zyvor.dev · info@zyvor.dev

/** Stable daemon `error_code` values → short user-facing labels. */
export const API_ERROR_LABELS: Record<string, string> = {
  operation_failed: 'The operation failed on the server',
  not_found: 'The requested resource was not found',
  invalid_request: 'The request was invalid',
  forbidden: 'You do not have permission for this action',
  libvirt_connection: 'Could not connect to libvirt on the host',
  internal_error: 'An internal server error occurred',
  unauthorized: 'Authentication required or your session expired',
  conflict: 'That name is already in use',
  invalid_name: 'Invalid name — use letters, numbers, spaces, and common punctuation',
}
const ERROR_CODE_RE =
  /\((operation_failed|not_found|invalid_request|forbidden|libvirt_connection|internal_error|unauthorized|conflict|invalid_name)\)\s*$/i

/** Human label for a stable API error code. */
export function friendlyErrorCode(code: string): string {
  return API_ERROR_LABELS[code] ?? code.replace(/_/g, ' ')
}

/** Strip HTML pages and trim noisy API text for UI display. */
export function sanitizeErrorText(text: string): string {
  const t = text.trim()
  if (!t) return ''

  if (/<!DOCTYPE\s+html/i.test(t) || /<html[\s>]/i.test(t) || t.includes('</html>')) {
    return (
      'The API returned an HTML error page instead of JSON. ' +
      'This usually means a reverse proxy failure or an OpenStack service (Keystone, Nova, Neutron, or Glance) is down.'
    )
  }

  if (t.includes('<') && t.includes('>')) {
    const stripped = t.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim()
    if (stripped.length > 0 && stripped.length < t.length * 0.6) {
      return stripped.length > 500 ? `${stripped.slice(0, 497)}…` : stripped
    }
  }

  return t.length > 600 ? `${t.slice(0, 597)}…` : t
}

/**
 * Build a user-facing message from an HTTP error body (JSON `{ error, error_code }` or plain/HTML).
 */
export function formatHttpErrorBody(status: number, statusText: string, text: string): string {
  const raw = text.trim()
  const statusLabel = `HTTP ${status}${statusText ? ` ${statusText}` : ''}`

  if (!raw) {
    return `Request failed (${statusLabel})`
  }

  if (/<!DOCTYPE\s+html/i.test(raw) || /^<\s*html/i.test(raw)) {
    return (
      `Request failed (${statusLabel}): the server returned an HTML error page instead of JSON. ` +
      'Check that machina-daemon is running and OpenStack endpoints are reachable.'
    )
  }

  try {
    const j = JSON.parse(raw) as { error?: string; message?: string; error_code?: string; remediation?: string }
    const code = typeof j.error_code === 'string' ? j.error_code : undefined
    const remediation = typeof j.remediation === 'string' ? j.remediation : undefined
    const rawMsg =
      typeof j.error === 'string' ? j.error : typeof j.message === 'string' ? j.message : ''
    const clean = sanitizeErrorText(rawMsg)

    if (clean) {
      if (code && (clean === code || clean === friendlyErrorCode(code))) {
        return remediation ? `${friendlyErrorCode(code)} — ${remediation}` : friendlyErrorCode(code)
      }
      if (code && !clean.toLowerCase().includes(code.replace(/_/g, ' '))) {
        const base = `${clean} (${friendlyErrorCode(code)})`
        return remediation ? `${base} — ${remediation}` : base
      }
      return remediation ? `${clean} — ${remediation}` : clean
    }
    if (code) {
      return friendlyErrorCode(code)
    }
  } catch {
    /* not JSON */
  }

  const sanitized = sanitizeErrorText(raw)
  if (sanitized !== raw) {
    return sanitized
  }
  if (raw.length > 400) {
    return `Request failed (${statusLabel}): ${raw.slice(0, 200)}…`
  }
  return `Request failed (${statusLabel}): ${raw}`
}

/** Format any thrown value for toasts and banners. */
/** True when the platform controller returned 404 / not_found for a VM or other resource. */
export function isPlatformNotFoundError(e: unknown): boolean {
  if (e && typeof e === 'object' && 'error_code' in e) {
    const code = (e as { error_code?: string }).error_code
    if (code === 'not_found') return true
  }
  const msg = formatUserError(e).toLowerCase()
  return /\bnot[_ ]found\b/.test(msg)
}

export function formatUserError(e: unknown): string {
  if (e instanceof Error) {
    const msg = sanitizeErrorText(e.message)
    const m = msg.match(ERROR_CODE_RE)
    if (m?.[1] && msg.replace(ERROR_CODE_RE, '').trim().length < 8) {
      return friendlyErrorCode(m[1])
    }
    return msg || 'Unknown error'
  }
  return sanitizeErrorText(String(e)) || 'Unknown error'
}

export type ParsedApiError = {
  message: string
  code?: string
}

export function parseUserError(e: unknown): ParsedApiError {
  const message = formatUserError(e)
  const m = message.match(ERROR_CODE_RE)
  return { message: message.replace(ERROR_CODE_RE, '').trim() || message, code: m?.[1] }
}

File: core/src/api_error.rs
// Copyright (c) 2026 ZyvorAI Labs Private Limited. All rights reserved.
// Proprietary software — see LICENSE in the repository root.
// https://zyvor.dev · info@zyvor.dev

//! User-facing HTTP/API error formatting (shared by daemon helpers and TUI).

use serde::Deserialize;

const API_ERROR_LABELS: &[(&str, &str)] = &[
    ("operation_failed", "The operation failed on the server"),
    ("not_found", "The requested resource was not found"),
    ("invalid_request", "The request was invalid"),
    ("forbidden", "You do not have permission for this action"),
    (
        "libvirt_connection",
        "Could not connect to libvirt on the host",
    ),
    ("internal_error", "An internal server error occurred"),
    (
        "unauthorized",
        "Authentication required or your session expired",
    ),
];

/// Human label for a stable API `error_code`.
pub fn friendly_error_code(code: &str) -> String {
    for (k, v) in API_ERROR_LABELS {
        if *k == code {
            return (*v).to_string();
        }
    }
    code.replace('_', " ")
}

/// Strip HTML pages and trim noisy API text for UI display.
pub fn sanitize_error_text(text: &str) -> String {
    let t = text.trim();
    if t.is_empty() {
        return String::new();
    }

    let lower = t.to_ascii_lowercase();
    if lower.contains("<!doctype html") || lower.contains("<html") || t.contains("</html>") {
        return "The API returned an HTML error page instead of JSON. \
            This usually means a reverse proxy failure or an OpenStack service \
            (Keystone, Nova, Neutron, or Glance) is down."
            .to_string();
    }

    if t.contains('<') && t.contains('>') {
        let stripped: String = t
            .split('<')
            .map(|seg| seg.split('>').nth(1).unwrap_or(seg))
            .collect::<Vec<_>>()
            .join(" ");
        let stripped = stripped.split_whitespace().collect::<Vec<_>>().join(" ");
        if !stripped.is_empty() && stripped.len() < t.len() * 6 / 10 {
            return truncate_plain(&stripped, 500);
        }
    }

    truncate_plain(t, 600)
}

fn truncate_plain(s: &str, max: usize) -> String {
    if s.chars().count() <= max {
        s.to_string()
    } else {
        let end = s.char_indices().nth(max).map(|(i, _)| i).unwrap_or(s.len());
        format!("{}…", &s[..end])
    }
}

#[derive(Debug, Deserialize)]
struct ApiErrorJson {
    error: Option<String>,
    message: Option<String>,
    error_code: Option<String>,
}

/// Build a user-facing message from an HTTP error body (JSON `{ error, error_code }` or plain/HTML).
pub fn format_http_error_body(status: u16, status_text: &str, text: &str) -> String {
    let raw = text.trim();
    let status_label = if status_text.is_empty() {
        format!("HTTP {status}")
    } else {
        format!("HTTP {status} {status_text}")
    };

    if raw.is_empty() {
        return format!("Request failed ({status_label})");
    }

    let lower = raw.to_ascii_lowercase();
    if lower.starts_with("<!doctype html") || lower.starts_with("<html") {
        return format!(
            "Request failed ({status_label}): the server returned an HTML error page instead of JSON. \
             Check that machina-daemon is running and OpenStack endpoints are reachable."
        );
    }

    if let Ok(j) = serde_json::from_str::<ApiErrorJson>(raw) {
        let code = j.error_code.as_deref();
        let raw_msg = j.error.as_deref().or(j.message.as_deref()).unwrap_or("");
        let clean = sanitize_error_text(raw_msg);

        if !clean.is_empty() {
            if let Some(c) = code {
                let label = friendly_error_code(c);
                if clean == c || clean == label {
                    return label;
                }
                if !clean.to_ascii_lowercase().contains(&c.replace('_', " ")) {
                    return format!("{clean} ({label})");
                }
            }
            return clean;
        }
        if let Some(c) = code {
            return friendly_error_code(c);
        }
    }

    let sanitized = sanitize_error_text(raw);
    if sanitized != raw {
        return sanitized;
    }
    if raw.len() > 400 {
        // Byte-index slicing panics if `raw` contains multi-byte UTF-8 and byte
        // 200 doesn't land on a char boundary (this text comes from an external
        // HTTP response body, so it's not guaranteed to be ASCII); truncate on a
        // char boundary instead, same as `truncate_plain` above.
        let end = raw.char_indices().nth(200).map(|(i, _)| i).unwrap_or(raw.len());
        return format!("Request failed ({status_label}): {}…", &raw[..end]);
    }
    format!("Request failed ({status_label}): {raw}")
}

/// Format any error string for status bars and toasts (strips HTML, maps bare error codes).
pub fn format_user_error(msg: &str) -> String {
    let msg = sanitize_error_text(msg);
    if msg.is_empty() {
        return "Unknown error".to_string();
    }
    for (code, label) in API_ERROR_LABELS {
        let suffix = format!("({code})");
        if msg.ends_with(&suffix) {
            let base = msg.trim_end_matches(&suffix).trim();
            if base.len() < 8 {
                return label.to_string();
            }
        }
    }
    msg
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn html_body_is_sanitized() {
        let body = "<!DOCTYPE html><html><body>Bad Gateway</body></html>";
        let msg = format_http_error_body(502, "Bad Gateway", body);
        assert!(!msg.contains("</html>"));
        assert!(msg.contains("HTML"));
    }

    #[test]
    fn json_error_code_only() {
        let body = r#"{"error_code":"operation_failed"}"#;
        let msg = format_http_error_body(500, "", body);
        assert!(msg.contains("operation failed"));
    }

    #[test]
    fn json_error_and_code() {
        let body = r#"{"error":"Neutron down","error_code":"operation_failed"}"#;
        let msg = format_http_error_body(503, "", body);
        assert!(msg.contains("Neutron"));
        assert!(msg.contains("operation failed"));
    }
}
